h/util/db.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import sqlalchemy
+
+
+def on_transaction_end(session):
+    """
+    Decorator for a function which should run after a top-level transaction ended.
+
+    Transaction that are either implicitly or explicity committed or rolled back will be
+    closed at the end of a Pyramid view. This is here for cleaning up caches so that
+    code after the view, exception views for example, will not be able to access
+    detached instances.
+
+    The wrapped function will receive the database session as ``session`` and the
+    transaction as ``transaction``.
+
+    Example usage:
+
+    .. code-block:: python
+
+       @util.db.on_transaction_end(session)
+       def flush_cache(session, transaction):
+           self._cache = {}
+
+    """
+    def wrapped(func):
+        @sqlalchemy.event.listens_for(session, 'after_transaction_end')
+        def after_transaction_end(current_session, transaction):
+            # We only clear the cache when the top-level transaction finishes.
+            if transaction.parent is None:
+                func(current_session, transaction)
```

Comment:
So this is a decorator, which means it should return the function that it's wrapping. In this particular case that's easy, because you don't actually need to modify the function at all (that's not we made it a decorator), so you can use [a similar pattern to the `listens_for` decorator itself](https://github.com/zzzeek/sqlalchemy/blob/c04870ba7b8098c7d408ad66f60efe7229496fde/lib/sqlalchemy/event/api.py#L92-L126).

So maybe something like:

```python
def on_transaction_end(session):
    def decorate(func):
        def _handler(_, transaction):
            if transaction.parent is None:
                func()
        sqlalchemy.event.listen(session, 'after_transaction_end', _handler)
        return func
    return decorate
```